src/java.base/share/classes/java/lang/reflect/Executable.java

```diff
@@ -311,11 +310,11 @@ public int getParameterCount() {
         // this case, we just return the result of
         // getParameterTypes().
         if (!genericInfo) {
-            return getParameterTypes();
+            return getSharedParameterTypes();
         } else {
             final boolean realParamData = hasRealParameterData();
             final Type[] genericParamTypes = getGenericParameterTypes();
-            final Type[] nonGenericParamTypes = getParameterTypes();
+            final Type[] nonGenericParamTypes = getSharedParameterTypes();
```

Comment:
(If removing these allocations is important, I note that `ConstructorRepository.getGenericParameterTypes()` lacks an equivalent to `getSharedParameterType` for trusted callers)

Similarly here I'm not sure the win from avoiding the clone is worth delegating the additional responsibility. You could still do this here if you add `.clone()` after `nonGenericParamTypes` on L344

Reply:
@cl4es Do you think it's worth creating a separate PR for `ConstructorRepository.getGenericParameterTypes()`?

Reply:
@stsypanov the remaining improvements in this PR are fine, simple wins. Adding complexity to expose shared arrays would warrant a bit more proof, however. For example a microbenchmark. My gut feeling is it's a marginal win on likely to be rare operations, but I couldn't know for sure without any evidence.